### ISSUE TITLE
Jerky playback with KWin compositing (strangely just with software decoding)

### ISSUE THREAD

**Issue by aufkrawall** [mpv-player/mpv#3964]
mpv 0.23 x64 on latest OpenSuse Tumbleweed

Problem:
When I play an H.264 or VP9 video with software decoding, playback looks jerky even though there are no dropped, repeated or delayed frames reported.
Playback looks normal when I turn off KWin compositing.
The weird thing is that it also looks normal when I use vdpau-copy and leave KWin compositing enabled.
Also games with vsync are running smoothly while keeping KWin compositing enabled, so it shouldn't be a general issue with my system config.
I tried out lots of mpv, KWin and Nvidia driver tweaks without success.

Expected behavior:
I would like to use KWin compositing with mpv without hardware decoding(e.g. because Nvidia currently doesn't offer VP9 decoding via VDPAU).
Turning off compositing is an ugly workaround, e.g. then lots of frames are dropped when I move windows, get tearing in the browser while mpv playback is going on  etc.
It works fine on Windows with DWM, btw.

Log file software decoding:
http://sprunge.us/KCBI 

log file vdpau-copy:
http://sprunge.us/RNid

**Comment by aufkrawall** [comment]
Ok, when watching really, really cose, I can also see some minor stutter with VDPAU-copy H.264.
So most likely not mpv's fault at all.

**Comment by kevinlekiller** [comment]
What I do is a bit complex, it might solve your issue,  I get no tearing and better performance, my `Xorg` and `kwin_x11` processes use significantly less CPU (both would use ~90% CPU usage previously, while moving a window around, now ~8%).

mpv: `vo=opengl`, `hwdec=cuda` (gives me better performance than vdpau), `video-sync=display-resample` [here's the whole mpv.conf if you're curious](https://gist.githubusercontent.com/kevinlekiller/374c8dd7139238e2c130bc1c5e757979/raw/ead03ea0976c0ff20a7dddefe76c4d91ea1dc186/gistfile1.txt)

kwin: Xrender backend, tearing prevention never

xorg.conf: `Option  "metamodes"     "nvidia-auto-select +0+0 { ForceFullCompositionPipeline = On }"` under `Section "Screen"`

When I want to switch refresh rates or resolutions I use a script which runs : `nvidia-settings --assign CurrentMetaMode="$RESOLUTION"_"$RATE +0+0 { ForceFullCompositionPipeline = On }"` where `$RESOLUTION`" is is the resolution and `$RATE` is the refresh rate.

I made a script called mpv in ~/bin which is in my users `$PATH`.

When I run mpv,  it runs this script which disables the kwin compositor (probably not required, but I used to do this in the past, so I just left it there) and sets the nvidia's GPU performance to its max performance state (because the driver on linux is junk and will stay in its lowest state while playing videos and cause the video to play poorly), the script is very long and does much more than that, so I modified it and [uploaded it here](https://gist.githubusercontent.com/kevinlekiller/faf138b5da012839f5619ab2a1bfa8bd/raw/90c27c6f3556a068dd9f69d6a3e2345bca82a6a9/gistfile1.txt).

**Comment by aufkrawall** [comment]
Thx. I noted as well that the adaptive power profile in the Nvidia driver on Linux is quite a desaster (even with newest 375.26 on a GTX 1070): It stays in faster states quite too long, but doesn't switch into them fast enough.
Thus I reproducibly get delayed frames if power profile isn't set to maximum performance. Well, since the power savings nearly don't exist anyway, I simply forced maximum performance globally all the time.
But good to know that there's also a script for mpv.

Regarding ForceFullCompositionPipeline/ForceCompositionPipeline:
That never worked for me without stutter (no matter how KWin is configured).
The only thing to ensure that there is absolutely zero stutter for me is to completely disable any global compositing and vsync and let every OpenGL application do its own vsync.
I verify this with good ol' Counter-Strike 1.6, Firefox (vsynctester.com with forced OGL rendering) and mpv with interpolation and some 60fps videos (e.g. Big Buck Bunny and juddertest videos).
I'm always on forced 75Hz custom resolution, btw.
In compton, applications don't seem to suffer by any stutter as well, but it's stuttery when moving windows or can't prevent tearing by 100%. So it doesn't make much sense to me to use it over plain X11 rendering.

Some time ago, some one else complained about problems with mpv and KWin:
https://forum.kde.org/viewtopic.php?f=111&t=128058
Seems this is still not sorted out, there is no config tweak that would eliminate the stutter completely.

**Comment by aufkrawall** [comment]
Weird. Here even moving windows is extremely stuttery with KWin Xrender (vsync off) + ForceFullCompositionPipeline. :(

**Comment by kevinlekiller** [comment]
Tried messing with the phone settings, but can't seem to get it to film higher than 20fps.

I've uploaded a video here, but it looks pretty bad because of the low fps from the camera, but in real life it's smooth: https://0x0.st/pJN.mp4

The youtube video playing is https://www.youtube.com/watch?v=5xkNy9gfKOg

**Comment by aufkrawall** [comment]
Hello again,
could you maybe give me a hint why cuda decoding doesn't work for me? It just silently falls back to software.
I'm using a mpv build provided by OpenSuse, it lists cuda as a possible option for hwdec. Is this not sufficient and I need a compiled build with cuda flag explicitly enabled?

**Comment by kevinlekiller** [comment]
Hi, I don't specifically enable a cuda flag when compiling, [these] (https://github.com/mpv-player/mpv/blob/7be495b369fa9924d0595a2ef0d2a0d1f662ba76/wscript#L871) are the checks when waf configure runs, and when you run waf configure you should see `Checking for CUDA hwaccel                                            : yes`.

As you can see the cuda.c file waf configure looks for is included with mpv's source : https://github.com/mpv-player/mpv/blob/master/waftools/fragments/cuda.c

It could be falling back to software decoding because the video itself is encoded in a way that the cuda decoder is not able to decode.

**Comment by wm4** [comment]
--hwdec always accepts the cuda value, even if cuda support is not compiled.

Cuda support depends on FFmpeg support too.

**Comment by aufkrawall** [comment]
Thanks, with a self compiled mpv build with latest ffmpeg git, cuda works.
Native cuda (hwdec=cuda) however doesn't work well for me with 60fps video, it looks like 30fps then.
Good news however is, cuda-copy fixes this. It even fixes the weird stutter with KWin this tracker entry was initially about. ;)
Nvidia also supports additional formats with cuda over vdpau, VP9 & HEVC 10 bit. Full of win.

This seems a good opportunity to describe my views and results on X11 compositor settings with Nvidia:
I'm using export __GL_YIELD="USLEEP" config tweak for KWin under "/home/.config/plasma-workspace/env/kwin_vars.sh" to eliminate the tearing problem of Nvidia driver.
Can't recommend setting up forced triple buffering in the driver/KWin since this introduces tearing or stutter problems in Firefox.
Can't usually recommend  setting up ForceCompositionPipeline in the Nvidia driver either, as this introduces (mouse) lag and potential for additional problems. 
But: It may be helpful if KWin compositing doesn't work for some reasons or you are with a different window manager. In this case, the combination of Compton + ForceCompositionPipeline can give a decent enough result. Compton just needs to be set up right with
backend = "xrender";
vsync = "none";
sw-opti = false;
unredir-if-possible = false;
(some are default settings, just pointing them out)
to have a tear-free, relatively fluid and stable compositing in conjuction with mpv.

Btw: Thanks, Kevin, for the power profile script for mpv + Nvidia driver. Extremely helpful.

**Comment by wm4** [comment]
> Native cuda (hwdec=cuda) however doesn't work well for me with 60fps video, it looks like 30fps then.

Probably too slow. Though I wouldn't know why.

**Comment by aufkrawall** [comment]
It seems the issue only affects the first seconds of an opened video and occurs quite randomly.
opengl-backend=x11egl also further reduces the chance compared to x11, it rarely ever happens with x11egl. x11egl also shows less CPU load than x11.
Probably it should select x11egl automatically when cuda is used?

**Comment by aufkrawall** [comment]
Just to keep you updated: The issue is caused by KWin when it compositing was disabled in fullscreen and you start windowed video playback after this.
KWin needs some time to "recover" when it starts with compositing anew, during this time it stutters. So, turning off compositing should be set to never in mpv's config to avoid this.

I also found out what's causing the weird stutter with software decoding which is not recognized by mpv's stats: It's Intel's powersave pstate CPU frequency governor (which is default for every distribution).
The stutter disappers when setting to intel_pstate performance or acpi-cpufreq performance governor.

So, to sum up: When you experience stuttering, disable downclocking of both your CPU and GPU, both can/will ruin the party.

However, this issue still exists apart from that:
https://github.com/mpv-player/mpv/issues/4003
